[JENKINS-75630] Add support for InfluxDB 3.0

g3n35i5 opened the pull request:
- Adding the `influxdb3-java` client library for InfluxDB v3 servers as runtime dependency
- Add support for InfluxDB 3.0
- 🚨 Remove field/tag duplications, see f67aa14 and 912bbcf for details.

This resolves [JENKINS-75630](https://issues.jenkins.io/browse/JENKINS-75630)

### Testing done

#### Existing automated tests

Adapted integration tests to test the InfluxDB v3 reporting as well.

<details>
<summary>InfluxDB v1.11</summary>

![Screenshot 2025-05-09 at 10 52 15](https://github.com/user-attachments/assets/cb22a8fd-a58e-48af-8fe3-5683436d8b93)
</details>

<details>
<summary>InfluxDB v2.7</summary>

![Screenshot 2025-05-09 at 10 52 29](https://github.com/user-attachments/assets/3a81260e-09bd-4b51-9bcf-aa69f64e4d55)
</details>

<details>
<summary>InfluxDB v3.2.0</summary>

![Screenshot 2025-05-09 at 10 52 46](https://github.com/user-attachments/assets/6696357c-d8f8-4615-90be-62a86f85fc94)
</details>

- Running several jobs, all with global listener enabled

<details>
<summary>Job Result</summary>

![Screenshot 2025-05-09 at 10 56 20](https://github.com/user-attachments/assets/10798ee7-123b-4a6c-91b4-f2dfe7aeea4b)
</details>

### Submitter checklist
- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira
- [x] Link to relevant pull requests, esp. upstream and downstream changes
- [x] Ensure you have provided tests - that demonstrates feature works or fixes the issue

<!--
Put an `x` into the [ ] to show you have filled the information.
The template comes from https://github.com/jenkinsci/.github/blob/master/.github/pull_request_template.md 
You can override it by creating .github/pull_request_template.md in your own repository 
-->

g3n35i5 committed:
build: Add `influxdb3-java` client library

g3n35i5 commented on src/main/java/jenkinsci/plugins/influxdb/models/Target.java:
```diff
@@ -201,56 +223,29 @@ public FormValidation doCheckDatabase(@QueryParameter String value) {
-                StandardUsernamePasswordCredentials credentials = CredentialsMatchers.firstOrNull(lookupCredentials, CredentialsMatchers.withId(null == credentialsId ? "" : credentialsId));
-
-                if (organization != null && !organization.trim().isEmpty()) {
-                    InfluxDBClientOptions.Builder options = InfluxDBClientOptions.builder().url(url).org(organization).bucket(database);
```
Here, no proxy is used. Did the proxy feature even work before? 🫣

g3n35i5 commented on src/main/java/jenkinsci/plugins/influxdb/InfluxDbPublicationService.java:
```diff
@@ -294,33 +320,7 @@ private void addPointsFromPlugin(List<Point> pointsToWrite, PointGenerator gener
-                    .url(target.getUrl())
-                    .org(target.getOrganization())
-                    .bucket(target.getDatabase())
-                    .okHttpClient(createHttpClient(url, target.isUsingJenkinsProxy()));
```
I reviewed my code again and I think I might broke the proxy feature. But this is the only section in the code that used the proxy, anyways. And in the `Target` class, no proxy was used. And in my understanding, this check was for the Jenkins settings (connection check).

asimell replied:
I just did a search in the repo about this. Looks like you're right. To my understanding, the v2 client uses proxy, but v1 client doesn't. It would be pretty simple to add the proxy support to the new client wrapper you introduce here at least for the v2 client.

This is one of those features that I haven't actually needed myself, it was implemented in a PR. There hasn't been any bug reports about this _not_ working, so it has gone under the radar until now. And a proxy setting is pretty difficult to test with unit tests as all endpoints are mocked anyway. Would you be able to fix the proxy config with this PR as well?

asimell commented:
Overall this PR looks very good! Great work so far!

You're right about the comments about proxy settings. I answered it more thoroughly directly in the comment. But as this is a thousand line PR, please understand that reviewing it will naturally take some time, and I need to setup a local InfluxDB 3.x client to be able to properly test this myself as well.

g3n35i5 commented:
> Overall this PR looks very good! Great work so far!
> 
> You're right about the comments about proxy settings. I answered it more thoroughly directly in the comment. But as this is a thousand line PR, please understand that reviewing it will naturally take some time, and I need to setup a local InfluxDB 3.x client to be able to properly test this myself as well.

Thank you for the positive feedback 😊

I've fixed/added the proxy support for V2.X servers in 5820b9b

g3n35i5 commented:
@asimell here's the docker compose I've used for testing

<details>

<summary>Docker Compose file for InfluxDB v1, v2 and v3 server</summary>

```yaml
services:
  influxdbv1:
    image: influxdb:1.11
    container_name: influxdbv1
    ports:
      - "8085:8086"

    environment:
      - INFLUXDB_ADMIN_ENABLED=true
      - INFLUXDB_ADMIN_USER=admin
      - INFLUXDB_ADMIN_PASSWORD=adminpassword
      - INFLUXDB_DB=prometheus
      - INFLUXDB_HTTP_LOG_ENABLED=true
      - INFLUXDB_REPORTING_DISABLED=true
      - INFLUXDB_USER=prometheus
      - INFLUXDB_USER_PASSWORD=prompass
    volumes:
      - influxdb-data-v1:/var/lib/influxdb
  influxdbv2:
    image: influxdb:2.7
    container_name: influxdbv2
    ports:
      - "8086:8086"
    volumes:
      - influxdb-data-v2:/var/lib/influxdb2
    environment:
      - DOCKER_INFLUXDB_INIT_MODE=setup
      - DOCKER_INFLUXDB_INIT_USERNAME=admin
      - DOCKER_INFLUXDB_INIT_PASSWORD=adminpassword
      - DOCKER_INFLUXDB_INIT_ORG=myorg
      - DOCKER_INFLUXDB_INIT_BUCKET=mybucket
      - DOCKER_INFLUXDB_INIT_ADMIN_TOKEN=mysecrettoken

  influxdb3-core:
    container_name: influxdb3-core
    image: influxdb:3-core
    ports:
      - 8181:8181
    command:
      - influxdb3
      - serve
      - --node-id=node0
      - --object-store=file
      - --data-dir=/var/lib/influxdb3

    volumes:
      - influxdb-data-v3:/var/lib/influxdb3

  grafana:
    image: grafana/grafana:latest
    container_name: grafana
    ports:
      - "3000:3000"
    volumes:
      - grafana-storage:/var/lib/grafana
    environment:
      - GF_SECURITY_ADMIN_USER=admin
      - GF_SECURITY_ADMIN_PASSWORD=admin

volumes:
  influxdb-data-v1:
  influxdb-data-v2:
  influxdb-data-v3:
  grafana-storage:
```
</details>

In order to create a database/token for the InfluxDB v3 server, start it and run these commands once:

```bash
docker exec -it influxdb3-core influxdb3 create token --admin
docker exec -it influxdb3-core influxdb3 create database jenkins --token <TOKEN>
```

g3n35i5 commented:
@asimell Do you have any ETA for your review? I know the PR is very large and if I can help to make it even simpler I can still make changes. However, I would be happy if the feature was available soon :blush:

asimell commented:
@g3n35i5 Hi, as a matter of fact I just started to look at this about 10 minutes ago. I'll try to get this reviewed and released next week.

asimell commented on src/main/java/jenkinsci/plugins/influxdb/models/InfluxDBClientWrapper.java:
```diff
@@ -0,0 +1,300 @@
+            boolean usingJenkinsProxy
+    ) {
+        // InfluxDB v2.X
+        if (organization != null && !organization.trim().isEmpty()) {
+            if (tryConnectV2(url, organization, database, basicAuthCredentials, tokenCredentials, usingJenkinsProxy)) {
+                return;
+            }
+        }
+        // InfluxDB v3.X/v1.X
+        else {
+            if (tryConnectV3(url, database, tokenCredentials)) {
+                return;
+            } else if (tryConnectV1(url, database, retentionPolicy, basicAuthCredentials)) {
+                return;
+            }
+        }
+        throw new RuntimeException("InfluxDB connection failed. Please check your connection parameters.");
```
There's some very serious bug in this logic. I added an InfluxDB 3 client to my local Jenkins, which also has an InfluxDB 1.8 target and an InfluxDB 2.x target configured. Connecting to InfluxDB 2 and InfluxDB 3 works, but testing InfluxDB 1.8 connection with username+password credentials, I simply get

```
java.lang.RuntimeException: InfluxDB connection failed. Please check your connection parameters.
	at PluginClassLoader for influxdb//jenkinsci.plugins.influxdb.models.InfluxDBClientWrapper.<init>(InfluxDBClientWrapper.java:79)
	at PluginClassLoader for influxdb//jenkinsci.plugins.influxdb.models.Target$DescriptorImpl.doVerifyConnection(Target.java:236)
```

And in the Jenkins logs I can see

```
2025-06-06 10:39:54.795+0000 [id=18]	WARNING	j.p.i.m.InfluxDBClientWrapper#tryConnectV3: Connection failed: Cannot invoke "org.jenkinsci.plugins.plaincredentials.StringCredentials.getSecret()" because "tokenCredentials" is null
```

Seems like `tryConnectV3` should not even be tried if `tokenCredentials` is `null`.

g3n35i5 replied:
I am a bit confused why this isn't working, the entire `tryConnectV3` method is a big try/catch block that catches all exceptions. The warning in the Jenkins log is a result of the failed `getSecret()` call, but the `tryConnectV1` method should still be executed. Is there a log message `Attempting connection to InfluxDB v1.X API` after the warning? Could you please provide the `FINE` level logs?

However, to be more verbose in logging, I've added [f58c095e83b4acf92f9778fb5d6c5679ac17b61c](https://github.com/jenkinsci/influxdb-plugin/pull/187/commits/f58c095e83b4acf92f9778fb5d6c5679ac17b61c).

asimell commented on src/main/java/jenkinsci/plugins/influxdb/models/InfluxDBClientWrapper.java:
```diff
@@ -0,0 +1,305 @@
+                logger.fine("Attempting connection without credentials");
+                this.v1v2client = InfluxDBClientFactory.createV1(url, "", "".toCharArray(), database, retentionPolicy);
+            }
+            if (this.v1v2client.ping() && this.getAPIVersion().startsWith("v1")) {
```
After some debugging I found that this is the root cause why the connection test failed. The v1 API version is `1.x.y` instead of `v1.x.y`. By taking the `v` out my local 1.8.10 started to work.

```suggestion
            if (this.v1v2client.ping() && this.getAPIVersion().startsWith("1")) {
```

g3n35i5 replied:
Fixed in 0590bad

g3n35i5 replied:
Fun fact: InfluxDB v1.11 returns `v1.X` instead of `1.X` 🫠 
Fixed in 8b13914

asimell commented on src/main/java/jenkinsci/plugins/influxdb/models/InfluxDBClientWrapper.java:
```diff
@@ -0,0 +1,305 @@
+
+    public void writePoints(List<AbstractPoint> pointsToWrite) {
+        if (this.v3client != null) {
+            List<com.influxdb.v3.client.Point> v3Points = pointsToWrite.stream()
+                    .map(AbstractPoint::getV3Point)
+                    .collect(Collectors.toList());
+            this.v3client.writePoints(v3Points);
```
I get some weird behaviour here. It's attempting to pass a field into a tag field. 1.x and 2.x work, so it has to be something to do with the new API and how it handles things or how things are stored in the  `AbstractPoint` object.

```
Error when executing always post condition:
com.influxdb.v3.client.InfluxDBApiHttpException: HTTP status code: 400; Message: invalid column type for column 'build_result', expected iox::column_type::tag, got iox::column_type::field::string
	at PluginClassLoader for influxdb//com.influxdb.v3.client.internal.RestClient.request(RestClient.java:216)
	at PluginClassLoader for influxdb//com.influxdb.v3.client.internal.InfluxDBClientImpl.writeData(InfluxDBClientImpl.java:316)
	at PluginClassLoader for influxdb//com.influxdb.v3.client.internal.InfluxDBClientImpl.writePoints(InfluxDBClientImpl.java:156)
	at PluginClassLoader for influxdb//com.influxdb.v3.client.internal.InfluxDBClientImpl.writePoints(InfluxDBClientImpl.java:151)
	at PluginClassLoader for influxdb//jenkinsci.plugins.influxdb.models.InfluxDBClientWrapper.writePoints(InfluxDBClientWrapper.java:142)
	at PluginClassLoader for influxdb//jenkinsci.plugins.influxdb.InfluxDbPublicationService.perform(InfluxDbPublicationService.java:292)
```

g3n35i5 replied:
As stated in the [InfluxDB v3 documentation](https://docs.influxdata.com/influxdb3/core/write-data/best-practices/schema-design/#do-not-use-duplicate-names-for-tags-and-fields), it is not allowed to have tags and fields with the same name.

In the [JenkinsBasePointGenerator](https://github.com/jenkinsci/influxdb-plugin/blob/49c509ed9227914e163494d6fd9305450aa845da/src/main/java/jenkinsci/plugins/influxdb/generators/JenkinsBasePointGenerator.java#L102-L118), multiple measurements (not only the build result) are reported as both, tags and fields.

I think we have two options: We decide to report these measurements either as field **or** as tag
- non breaking: only for InfluxDB v3
- breaking: for InfluxDB v1 and v2 as well

Both options have their pros and cons. I'd prefer to report the exact same data to all InfluxDB instances, but this would probably break many queries...
```java
AbstractPoint point = buildPoint(measurementName, customPrefix, build);

        point.addField(BUILD_TIME, build.getDuration() == 0 ? dt : build.getDuration())
                ...
                .addField(BUILD_RESULT, result)
                ...
                .addTag(BUILD_RESULT, result);
```

g3n35i5 replied:
I decided to remove the tags/fields duplication in 5bc11ebfe42f47452e538145a1416e8e8d199047. Is this okay for you?

g3n35i5 commented:
@asimell I'll add some integration tests using docker to the pipeline to ensure everything is working as expected :+1:

Until then, I'll mark the PR as draft again.

g3n35i5 commented:
@asimell I get the [following build error](https://ci.jenkins.io/job/Plugins/job/influxdb-plugin/job/PR-187/11/pipeline-overview/)

```
Could not find a valid Docker environment. Please see logs and check configuration
```

I've set `useContainerAgent: false, // Set to `false` if you need to use Docker for containerized tests` in the Jenkinsfile, however I still don't have docker enabled.

Maybe this is caused by the early warning

```
‘Jenkinsfile’ has been modified in an untrusted revision
```

What do I need to do to enable docker usage?

**Update**
@asimell I think the cause is that my changes to the `Jenkinsfile` are ignored [for security reasons](https://docs.cloudbees.com/docs/cloudbees-ci-kb/latest/client-and-managed-controllers/why-jenkinsfile-changes-are-not-reflected-in-pr-build#_resolution) because I am not a maintainer resp. my fork isn't a trusted repo.

Could you make a PR where you set the `useContainerAgent` to false and check whether your pipeline still runs as expected? 😊 Then I'll rebase my PR to the new Jenkinsfile.

asimell commented:
I opened #188 to set container agents to False. Let's see how that runs and I'll merge if everything passes.

asimell commented:
#188 is now merged.

g3n35i5 commented:
I'll do the refactorings for InfluxDB v1/v2 (wrapper classes, etc.) including first integration tests in a separate PR(https://github.com/jenkinsci/influxdb-plugin/pull/189).

When this is working & merged, I'll rebase this PR.

g3n35i5 committed:
feat: Support InfluxDB v3

g3n35i5 committed:
refactor(tests): Improve check for expected records in integration test

g3n35i5 committed:
fix!: Remove duplicted fields/tags

- AbstractPointGenerator
  - Remove fields `PROJECT_NAME` and `PROJECT_PATH`. They already exist as tags.
- JenkinsBasePointGenerator
  - Remove field `BUILD_RESULT`. It already exists as tag.

asimell commented:
#189 is now merged. This can be converted from a draft to an actual PR. There also seems to be some merge conflicts which need to be fixed.

asimell commented on src/main/java/jenkinsci/plugins/influxdb/generators/AbstractPointGenerator.java:
```diff
@@ -41,14 +41,16 @@ public AbstractPointGenerator(Run<?, ?> build, TaskListener listener, ProjectNam
         String projectName = projectNameRenderer.render(build);
         String projectPath = build.getParent().getRelativeNameFrom(instance);
         AbstractPoint point = new AbstractPoint(name)
-                .addField(PROJECT_NAME, projectName)
-                .addField(PROJECT_PATH, projectPath)
```
Please add mentions of this to the [Available metrics](https://github.com/jenkinsci/influxdb-plugin/blob/development/doc/available_metrics.md) and [Breaking changes](https://github.com/jenkinsci/influxdb-plugin/blob/development/doc/breaking_changes.md) documentation. Since this is backwards incompatible, the next release will be a major release (`6.0`). See for example `rf_results` in `available_metrics.md`.

g3n35i5 replied:
Done in 894e13b

asimell commented on src/main/java/jenkinsci/plugins/influxdb/generators/JenkinsBasePointGenerator.java:
```diff
@@ -104,7 +104,6 @@ public AbstractPoint[] generate() {
                 .addField(BUILD_EXEC_TIME, build.getStartTimeInMillis())
                 .addField(BUILD_MEASURED_TIME, currTime)
                 .addField(BUILD_STATUS_MESSAGE, build.getBuildStatusSummary().message)
-                .addField(BUILD_RESULT, result)
```
Please add mentions of this to the [Available metrics](https://github.com/jenkinsci/influxdb-plugin/blob/development/doc/available_metrics.md) and [Breaking changes](https://github.com/jenkinsci/influxdb-plugin/blob/development/doc/breaking_changes.md) documentation. Since this is backwards incompatible, the next release will be a major release (`6.0`). See for example `rf_results` in `available_metrics.md`.

g3n35i5 replied:
Done in 894e13b

g3n35i5 committed:
docs: Update available metrics and breaking changes docs

- Fields `project_name` and `project_path` have been removed from **all metrics** since they are already defined as tags.
- Field `build_result` has been removed from the **jenkins_data** metrics since it is already defined as tag.

g3n35i5 commented:
@asimell I don't want to be intrusive, I'm sure you have a lot of other things on your plate. However, I would be happy if this feature and version 6.0 could be released soon, as I would really like to use this feature. Can I support you in any way, maybe with other PRs? Cheers!

asimell commented:
Hi @g3n35i5,

Apologies that this has taken a while. July is a typical holiday month in Finland and I haven't been at a computer that many times. I'm also currently on parental leave, so even when holidays are over I'm not that consitently on my computer 😅 and finally, it's my wife's and other child's final holiday week this week before they go to work and day care respectively 😅

So, unfortunately you'll have to wait to at least next week before I get proper focus time to review, merge, and release this. Sorry for the inconvenience.

g3n35i5 commented:
Hi @asimell,

Thank you for your response and no worries at all. I completely understand and appreciate the time you take to review and merge the PR. Enjoy your well-deserved time with your family, and whenever you get a chance to look at it, that would be great.
Thanks again and cheers!

asimell commented:
While testing I found out that this doesn't work with (at least) `junit_data`. There's still a duplication between tags and fields, which InfluxDB 3.x doesn't allow (?). Either the field or the tag should be deleted (?) and the corresponding documentation updated in `available_metrics.md` and `breaking_changes.md`.

These are the duplications I found:
* `junit_data`
    * `pipeline_step`
    * `test_status`
    * `suite_name`
    * `test_name`
    * `test_class_full_name`
 * `perfpublisher_test`
    * `test_name`
 * `perfpublisher_test_metric`
    * `test_name`
 * `testcase_point`
    * `rf_name`
 * `tag_point`
    * `rf_tag_name`
 * `suite_result`
    * `rf_suite_name`

g3n35i5 committed:
feat!: Remove more tag/field duplications

Removes the following tags from the following generators:

* `junit_data`
    * `pipeline_step`
    * `test_status`
    * `suite_name`
    * `test_name`
    * `test_class_full_name`
* `perfpublisher_test`
    * `test_name`
* `perfpublisher_test_metric`
    * `test_name`
* `testcase_point`
    * `rf_name`
* `tag_point`
    * `rf_tag_name`
* `suite_result`
    * `rf_suite_name`

g3n35i5 commented:
> While testing I found out that this doesn't work with (at least) `junit_data`. There's still a duplication between tags and fields, which InfluxDB 3.x doesn't allow (?). Either the field or the tag should be deleted (?) and the corresponding documentation updated in `available_metrics.md` and `breaking_changes.md`.
> 
> These are the duplications I found:
>     ...

@asimell thanks for spotting this one! Fixed in 912bbcf1470fbe347af5ee433b2052324eedf47b.

**Note: I decided to remove the tags because, as I understand it, fields are intended for data and tags are intended for metadata.**

asimell reviewed (approved):
✨ LGTM! ✨ 

Works like a charm. Apologies it took so long, I had a duplicate `build_number` tag in my own Jenkinsfile I used for testing, which caused the plugin to fail. Debugging this in the current life state was really demotivating, so I just kinda pushed it back.

g3n35i5 commented:
> ✨ LGTM! ✨
> 
> Works like a charm. Apologies it took so long, I had a duplicate `build_number` tag in my own Jenkinsfile I used for testing, which caused the plugin to fail. Debugging this in the current life state was really demotivating, so I just kinda pushed it back.

No worries, thanks for the review! Glad I could contribute to this project, looking forward for 6.0 :rocket:

g3n35i5 commented:
@asimell do you have any ETA for the 6.0 release? Cheers!

asimell commented:
@g3n35i5 6.0 has been released
